### PR TITLE
[REF] web, *: remove patchMixin + imp utils.patch

### PR DESCRIPTION
**Pull request by mcm-odoo**
* hr, hr_holidays, im_livechat, mail, snailmail, website,
  website_livechat

This commit removes `patchMixin` and improve `utils.patch`.
`utils.patch` now supports native classes and has a new parameter
used to patch class members.

`utils.patch` is now used everywhere `patchMixin` was and it must
be used to patch classes.

Enterprise PR: https://github.com/odoo/enterprise/pull/16278

### CODE REVIEW

**Comment by seb-odoo** [comment]
Why the change? I'm not against it, just curious why we made patchMixin the recommended way a few months ago and we are now backtracking?

The most annoying point I had with it was that it changed the actual class that was returned, and in debug or log I saw PatchMixin instead of actual class name everywhere.

**Comment by ged-odoo** [comment]
> Why the change? I'm not against it, just curious why we made patchMixin the recommended way a few months ago and we are now backtracking?
> 
> The most annoying point I had with it was that it changed the actual class that was returned, and in debug or log I saw PatchMixin instead of actual class name everywhere.

yeah, that was part of the problem.  At first, we made the PatchMixin to make it work with native class and native syntax (goal was to be able to use "super.someMethod".  But then, it made everything slightly more complicated, so we decided that the unified syntax was not worth it.  With the patch method, we have to use `_super`, and that's ok. At least, its more explicit

**Comment by mcm-odoo** [comment]
Hello @seb-odoo, as you mention it there were some annoying points with `patchMixin`.
There are some with `utils.patch` too but the main point is that we cannot patch any class with `patchMixin` at any time.
`patchMixin` needed two steps:
- Modules must return a "patch-able" version of a class to be able to patch it (and its sub classes must be "patch-able" too if we want to patch them).
- we patch the class that needs to be patched.

The advantage of `utils.patch` is that we have only 1 step, we don't need to tell the class to be "patch-able" we can patch at any time.
Also the implementation and the use of `patchMixin.patch` is a bit complicated, we need to pass a callback like `(T) => class extends T {...}`, with `utils.patch` we pass an object which is easier to write and understand.

File: addons/web/static/tests/fields/relational_fields/field_one2many_tests.js [hunk 1]
```diff
@@ -9595,24 +9596,22 @@ QUnit.module('fields', {}, function () {
...
-                    }
-                    async update() {
-                        // the issue is a race condition, so we manually delay the update to turn it deterministic
-                        await prom;
-                        super.update(...arguments);
-                    }
-                }
-            );
+            const unpatch = patch(ControlPanel, 'cp_patch_mock', {
+                setup() {
```
Comment (caburj) [inline addons/web/static/tests/fields/relational_fields/field_one2many_tests.js:9600]: The constructor from the old code is replaced by this `setup` method. But I can't find any call to this method. I'm expecting that it is called in the `constructor` of the patched class, but the patched class's constructor isn't modified to call `setup`. Do patched classes automatically call `setup` during instantiation?

(Sorry to just come out of the blue. I'm just very excited to this newly improved feature of patching. :) )

Comment (ged-odoo) [reply addons/web/static/tests/fields/relational_fields/field_one2many_tests.js:9600]: it's now always done for component by Owl itself: https://github.com/odoo/owl/commit/19a47a7001a90ca6fd6a1a668b7107e5af4c74bf (you need the last version of owl for that).  With this, all components can be patched easily.  In odoo, other classes can define a `setup` method if necessary.

Comment (caburj) [reply addons/web/static/tests/fields/relational_fields/field_one2many_tests.js:9600]: I see. Thanks.

**Review by ged-odoo — COMMENTED** [review]
missing doc. we need to document everything that we do, in particular if it is meant to be used by other people

File: addons/web/static/src/js/core/utils.js [hunk 2]
```diff
@@ -614,51 +614,85 @@ var utils = {
...
-                    if (!(methodName in metadata.origMethods)) {
-                        metadata.origMethods[methodName] = original;
+        if (staticPatch) {
+            applyPatch(C, staticPatch, CInfo.originalStatic);
+        }
+
+        function makeIntermediateFunction(key, prevDesc, newDesc) {
+            const _superFn = prevDesc[key];
+            const patchFn = newDesc[key];
+            newDesc[key] = function intermediatePatch(...args) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/utils.js:651]: `__patched__function__`

File: addons/web/static/src/js/core/utils.js [hunk 3]
```diff
@@ -649,8 +649,10 @@ var utils = {
             }
 
             if (prevDesc) {
+                const patchedFnName = `(patch) ${patchName}:${k}`;
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/utils.js:652]: `${k} (patch ${patchname})`

**Comment by ged-odoo** [comment]
robodoo r+

**Comment by ged-odoo** [comment]
robodoo r+

**Pull request closed without merging**